### PR TITLE
Enhc: Filter regimen line by age(Adult regimen line > age 14 < Child Regimen line)

### PR DESCRIPTION
**Pull request by Justus446**
## Requirements

- [x] This PR has a title that briefly describes the work done, including the ticket number if there is a ticket.
- [x] My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide).
- [ ] I checked for feature overlap with [**existing widgets**](https://om.rs/directory).

## Summary
**This PR create 2 hooks and enhances the edit regimen functionality :**
1.useCalculateAge:to get the age of a patient(The usePatient hook does not come with age precalculated)
2. useFilteredRegimen: to filter regimens based on age(adult regimens for age > 14, and child regimens for age< or = 14)
3. Drop downs to select regimens for standard-regimen-component and non-standard-regimen-component populated with the relevant regimens based on age

Avoid users selecting child regimens for adults and vice versa by displaying the correct regimen based on age

## Screenshots
image

*None.*

## Related Issue

*None.*

## Other

*None.*

### CODE REVIEW

File: packages/esm-care-panel-app/src/hooks/useCalculateAge.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      return age;
+    };
+
+    setAge(calculateAge(birthDate));
+  }, [birthDate]);
+
+  return age;
+}
+
+export default useCalculateAge;
```
Comment (donaldkibet) [inline packages/esm-care-panel-app/src/hooks/useCalculateAge.ts:32]: we can convert this to a function that calculate age as opposed to using a hook. UseEffect will trigger another re-render.

File: packages/esm-care-panel-app/src/hooks/useFilteredRegimen.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+interface RegimenItem {
+  name: string;
+  conceptRef: string;
+}
+
+interface RegimenLineGroup {
+  regimenline: string;
+  regimenLineValue: string;
+  regimen: RegimenItem[];
+}
```
Comment (donaldkibet) [inline packages/esm-care-panel-app/src/hooks/useFilteredRegimen.ts:12]: This should be in the types file

**Review by donaldkibet — CHANGES_REQUESTED** [review]
Thanks @Justus446 Great enhancement. Left some suggestions. Would also be good if this would be accompanied by Unit test to avoid regression. Thanks!

File: packages/esm-care-panel-app/src/regimen-editor/non-standard-regimen.component.tsx [hunk 3]
```diff
@@ -26,6 +29,9 @@ const NonStandardRegimen: React.FC<NonStandardRegimenProps> = ({
   const matchingCategory = standardRegimen.find((item) => item.categoryCode === 'ARV'); // Non standard regimen will be exclusively for ARVs
   const [selectedRegimens, setSelectedRegimens] = useState(Array(5).fill(''));
   const [nonStandardRegimenObjects, setStandardRegimenObjects] = useState([]);
+  const { patient } = usePatient();
+  const patientAge = useCalculateAge(patient?.birthDate);
+  const filteredRegimenByAge = useFilteredRegimen(matchingCategory?.category, patientAge);
```
Comment (makombe) [inline packages/esm-care-panel-app/src/regimen-editor/non-standard-regimen.component.tsx:34]:
```suggestion
  const filteredRegimenLineByAge = useFilteredRegimen(matchingCategory?.category, patientAge);
```

File: packages/esm-care-panel-app/src/regimen-editor/non-standard-regimen.component.tsx [hunk 4]
```diff
@@ -77,7 +83,7 @@ const NonStandardRegimen: React.FC<NonStandardRegimenProps> = ({
             {!selectedRegimenLine || selectedRegimenLine == '--' ? (
               <SelectItem text={t('selectRegimenLine', 'Select Regimen Line')} value="" />
             ) : null}
-            {matchingCategory?.category.map((line) => (
+            {filteredRegimenByAge.map((line) => (
```
Comment (makombe) [inline packages/esm-care-panel-app/src/regimen-editor/non-standard-regimen.component.tsx:86]:
```suggestion
            {filteredRegimenLineByAge.map((line) => (
```

File: packages/esm-care-panel-app/src/regimen-editor/standard-regimen.component.tsx [hunk 5]
```diff
@@ -61,7 +67,7 @@ const StandardRegimen: React.FC<StandardRegimenProps> = ({
             {!selectedRegimenLine || selectedRegimenLine == '--' ? (
               <SelectItem text={t('selectRegimenLine', 'Select Regimen Line')} value="" />
             ) : null}
-            {matchingCategory?.category.map((line) => (
+            {filteredRegimenByAge.map((line) => (
```
Comment (makombe) [inline packages/esm-care-panel-app/src/regimen-editor/standard-regimen.component.tsx:70]:
```suggestion
            {filteredRegimenLIneByAge.map((line) => (
```

File: packages/esm-care-panel-app/src/regimen-editor/standard-regimen.component.tsx [hunk 6]
```diff
@@ -24,6 +27,9 @@ const StandardRegimen: React.FC<StandardRegimenProps> = ({
   const [selectedRegimen, setSelectedRegimen] = useState('');
   const [selectedRegimens, setSelectedRegimens] = useState([]);
   const matchingCategory = standardRegimen.find((item) => item.categoryCode === category);
+  const { patient } = usePatient();
+  const patientAge = useCalculateAge(patient?.birthDate);
+  const filteredRegimenByAge = useFilteredRegimen(matchingCategory?.category, patientAge);
```
Comment (makombe) [inline packages/esm-care-panel-app/src/regimen-editor/standard-regimen.component.tsx:32]:
```suggestion
  const filteredRegimenLineByAge = useFilteredRegimen(matchingCategory?.category, patientAge);
```

File: packages/esm-care-panel-app/src/hooks/useCalculateAge.ts (hunk 1 above)
Comment (Justus446) [reply packages/esm-care-panel-app/src/hooks/useCalculateAge.ts:32]: useCalculateAge converted to a hook

File: packages/esm-care-panel-app/src/hooks/useFilteredRegimen.ts (hunk 2 above)
Comment (Justus446) [reply packages/esm-care-panel-app/src/hooks/useFilteredRegimen.ts:12]: interfaces moved to types file

File: packages/esm-care-panel-app/src/regimen-editor/non-standard-regimen.component.tsx (hunk 3 above)
Comment (Justus446) [reply packages/esm-care-panel-app/src/regimen-editor/non-standard-regimen.component.tsx:34]: filteredRegimenByAge renamed to filteredRegimenLineByAge

File: packages/esm-care-panel-app/src/regimen-editor/standard-regimen.component.tsx (hunk 5 above)
Comment (Justus446) [reply packages/esm-care-panel-app/src/regimen-editor/standard-regimen.component.tsx:70]: filteredRegimenByAge renamed to filteredRegimenLineByAge

File: packages/esm-care-panel-app/src/regimen-editor/standard-regimen.component.tsx (hunk 6 above)
Comment (Justus446) [reply packages/esm-care-panel-app/src/regimen-editor/standard-regimen.component.tsx:32]: filteredRegimenByAge renamed to filteredRegimenLineByAge

File: packages/esm-care-panel-app/src/regimen-editor/non-standard-regimen.component.tsx (hunk 4 above)
Comment (Justus446) [reply packages/esm-care-panel-app/src/regimen-editor/non-standard-regimen.component.tsx:86]: filteredRegimenByAge renamed to filteredRegimenLineByAge

**Comment by Justus446** [comment]
> Thanks @Justus446 Great enhancement. Left some suggestions. Would also be good if this would be accompanied by Unit test to avoid regression. Thanks!

changes made,unit tests coming in another PR

**Comment by makombe** [comment]
@Justus446  one more thing, after changing the methods to functions, they are no longer hooks and should be removed from the hooks folder into the util file. Thanks

**Comment by Justus446** [comment]
> @Justus446 one more thing, after changing the hooks to functions, they are no longer hooks and should be removed from the hooks folder into the `utils.tsx` file. Thanks

Done

**Review by makombe — APPROVED** [review]
LGTM

**Pull request merged**